language/bytecode-verifier/src/dependencies.rs

```diff
@@ -343,3 +384,74 @@ fn compare_structs(
         Ok(())
     }
 }
+
+fn verify_all_script_visibility_usage(context: &Context) -> PartialVMResult<()> {
+    match &context.resolver {
+        BinaryIndexedView::Module(m) => {
+            for (idx, fdef) in m.function_defs().iter().enumerate() {
+                let code = match &fdef.code {
+                    None => continue,
+                    Some(code) => &code.code,
+                };
+                verify_script_visibility_usage(
+                    context,
+                    fdef.visibility,
+                    FunctionDefinitionIndex(idx as TableIndex),
+                    code,
+                )?
+            }
+            Ok(())
+        }
+        BinaryIndexedView::Script(s) => verify_script_visibility_usage(
+            context,
+            Visibility::Script,
+            FunctionDefinitionIndex(0),
+            &s.code().code,
+        ),
+    }
+}
+
+fn verify_script_visibility_usage(
+    context: &Context,
+    current_visibility: Visibility,
+    fdef_idx: FunctionDefinitionIndex,
+    code: &[Bytecode],
+) -> PartialVMResult<()> {
+    macro_rules! invalid_script_visibile_call {
+        ($idx:expr) => {
+            Err(PartialVMError::new(StatusCode::CALLED_SCRIPT_VISIBLE_FROM_NON_SCRIPT_VISIBLE).at_code_offset(fdef_idx, $idx).with_message("script-visible functions can only be called from scripts or other script-visibile functions".to_string()))
+        }
+    }
+
+    for (idx, instr) in code.iter().enumerate() {
+        let idx = idx as CodeOffset;
+        match instr {
```

Comment:
Maybe (did not test it so it may not compile)

```rust
let fhandle_idx = match instr {
  Bytecode::Call(fhandle_idx) => fhandle_idx,
  Bytecode::CallGeneric(finst_idx) => &context.resolver.function_instantiation_at(*finst_idx).handle,
  _ => continue,
};
let fhandle_vis = context.function_visibilities[fhandle_idx];
match (current_visibility, fhandle_vis) {
  (Visibility::Script, Visibility::Script) => (),
  (_, Visibility::Script) => {
    return invalid_script_visibile_call!(idx);
  }
  _ => (),
}
```

Reply:
I like it